### DISCUSSION TITLE
Feedback: New to XAML Today

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/microsoft-ui-xaml#9741] [General]
I have been learning WinUI app development using XAML.

For context, my comments are from the perspective of a beginner in terms of Windows app development. My experiments were using the VS 2022 template for a blank app, packaged with the WAP project included.

<details>
Why WinUI 3 and not something else? Because information online indicates that WinUI 3 is the only Microsoft presentation framework that uses modern graphics drivers and hardware. 
</details>

First, using XAML, it is difficult to find the properties, methods, and events of controls without extensive use of learn.microsoft.com docs. However, each control sometimes uses similar terminology with differing meaning and impact. Hypothetically, the 'Color' property in one control may be the foreground color, while being used as the background color in a different control.

The docs also seem to better explain the properties, methods, and events from a C# perspective, while early remarks and examples use XAML. Perhaps a difference between handcrafted and auto-generated code documentation?

Searching for XAML help online is difficult. Years of changes and multiple frameworks using the same jargon (but different) make it hard to differentiate XAML information specific to WinUI 3. Even namespaces are confusingly similar.

Modifying XAML with a running application and hot reload did not modify the visual presentation on the desktop. It was necessary to close the application and rerun to see the changes. It would be helpful if there was a way to refresh the XAML code without shutting down the app and redeploying/restarting. That might make XAML useful.

<details>
My assumption was that the purpose of the XAML code was to avoid the need to reroll and redeploy the application--similar to how XML definitions files could allow a program to run without a restart and recompile simply by zeroing and re-initializing the XML defs while the program continues to run.  
</details>

As a new WinUI 3 developer, it is difficult for me to understand the "Why XAML" now with WinUI 3. I do understand that other presentation frameworks used XAML. but current documentation continues to encourage XAML as the best practice.

That said, given no visible benefit of XAML and with the VS 2022 C# IntelliCode help, I am struggling against my instincts to keep using XAML. Instead, I would like to use all C# code for all view work and ignore XAML entirely if possible.

<details>
I am familiar with HTML and some template engines for the web, but XAML seems too foreign to make that experience helpful.   
</details>

Thanks.

**Comment by eduardobragaxz** [comment]
My experience was that XAML at one point started making sense. Might just be that I'm used to it now. When searching for solutions make sure to add either "uwp" or "winui" in the query.

Hot Reload on XAML can be finicky sometimes, but it does work. I did notice, however, that setting the x:Load property makes them not change. Don't know if it's a bug.

**Reply by ghost** [reply]
> My experience was that XAML at one point started making sense.

It makes sense in terms of its own syntax. It does not sell me on its added value.

XAML does not simplify the order and visualization of a presentation view anymore than C# code would or does in my mind. It seems like just another syntax to learn without benefit. Why is XAML written WinUI 3 better than C# written WinUI 3?

Also, in terms of documentation, I've found that using "WinUI" adds to the mess in terms of search and research.

Using the namespace "Microsoft.UI.Xaml.Controls" as a search term is the only somewhat reliable way to get close to WinUI 3 documentation on learn.microsoft.com.

If I google WinUI Grid for example, the result is the Grid documentation for UWP and Windows.UI.Xaml.Controls.

Almost all searches for WinUI first indicate UWP documentation pages on learn.microsoft.com. The content on such learn pages look identical to other pages describing other XAML technologies. For example, that "Grid" page on both namespaces is only slightly different. That must not be helping search engines properly catalog.

Finally, hitting the correct namespace in learn, the version drop down is the "Windows App SDK 1.5" (as of now). WinUI 3 is never mentioned. Beware, a different namespace uses WinUI 2.x versions as the drop down and that further muddies the waters.

I can only take what I find at face value.

EDIT; Something else that really annoys. I create that base template project(s). It generates two XAML files each with an associated C# file.

If I create a "src" directory and move those files there the entire solution fails with errors.

If I only move the main window files into a "views" directory, the entire solution fails with errors.

Moving the files back never fixes these issues and I must delete and restart a new solution.

I did adjust the namespaces, class names, etc. in the XAML files.

Online research suggested many hacks using "startinguri" etc. None worked.

I think this should be something very easy to do no matter what else was engineered by conventions. If necessary to declare, it should be the same level of effort of a ubiquitous include directive. It should not be difficult or secret.

**Reply by eduardobragaxz** [reply]
I think I would hate something like SwiftUI. I prefer a different syntax.

**Reply by ghost** [reply]
> SwiftUI

I had to google SwiftUI to understand your comment. I assume you are implying C# looks like SwiftUI syntax.

I already "know" C# syntax. I also don't need apple.

Why learn anything unless it accelerates you more from a need to a working program?

For something to be worth using, it can never be more work than using C. Otherwise, why bother with that general purpose scripting at all? This is just like C makes getting to the run program simpler than assembly.

**Reply by eduardobragaxz** [reply]
I wasn't implying their syntax look alike, just that I don't think I would like to use C# syntax for UI. I like the separation between C# and XAML.

**Comment by ghost** [comment]
As an aside, given the XAML online documentation situation, what is the impact on AI responses and example code to questions regarding XAML? Something to consider as well when using XAML and AI assistance.

**Comment by yourordinarycat** [comment]
> First, using XAML, it is difficult to find the properties, methods, and events of controls without extensive use of learn.microsoft.com docs

The designer was quite useful for this, sadly WinUI 3 doesn't have one. I agree that this could use some improvement, adding property docs to the XAML Intellisense would be wonderful too.

> However, each control sometimes uses similar terminology with differing meaning and impact. Hypothetically, the 'Color' property in one control may be the foreground color, while being used as the background color in a different control

I haven't found this to be an issue - a more concrete example could be nice.

> Searching for XAML help online is difficult. Years of changes and multiple frameworks using the same jargon (but different) make it hard to differentiate XAML information specific to WinUI 3

I agree that the UWP+WinUI 2 and WinUI 3 documentation split was handled very badly, and needs improvement. But only those two, WPF documentation was never really an issue.

> As an aside, given the XAML online documentation situation, what is the impact on AI responses and example code to questions regarding XAML?

My guess is, about as much of a problem as "How to show a button using Swift?" would be on Apple platforms. So probably none, unless you're blindly trusting the LLM or can't tell the difference between UIKit and SwiftUI.

> Modifying XAML with a running application and hot reload did not modify the visual presentation on the desktop. It was necessary to close the application and rerun to see the changes

Visual Studio usually tells you why hot reload is failing - sometimes it fails for seemingly no reason though, it needs improvement, I agree. A designer would fix this entirely, and you can use one with UWP ;)

> If I create a "src" directory and move those files there the entire solution fails with errors

Are you moving them from within Visual Studio, or the explorer? If you're doing it the second way, that won't properly update the project files - XAML files in particular go through the XAML compiler task, so they're explicitly referenced in your .csproj, which Visual Studio will update on its own, but won't work if done from the explorer. Same applies if you're moving the whole project directory, because the solution file references those.

> As a new WinUI 3 developer, it is difficult for me to understand the "Why XAML" now with WinUI 3

Some of the technical aspects & benefits are pretty nicely explained here, in my opinion: https://github.com/microsoft/microsoft-ui-xaml/issues/4127#issuecomment-747668811

Subjectively, I'd argue writing XAML is a much better experience. You can already create your window and controls in C# if you really want to, but C# isn't designed to do UI, and WinUI isn't designed to be used without XAML. Example from that same thread, and not even using more advanced features like attached properties, events, behaviors, or design time data:

```XAML
<A b="c" d="e">
  <B f="g" h="i">
    <C j="k" l="m"></C>
  </B>
</A>
```

```C#
  A a = new A() { b: "c", d: "e", _: new B()
                  { f: "g", h: "i", _: new C()
                    { j: "k", l: "m"
                    }
                  }
                }
```
